### PR TITLE
Junit5 features

### PR DESCRIPTION
**Pull request by zwt299**
[Attempted implementation of timeout. Please check over...](https://github.com/EngineeringSoftware/inlinetest/pull/3)

### CODE REVIEW

File: python/tests/test_conftest.py [hunk 1]
```diff
@@ -383,3 +383,18 @@ def m(a, b, c, d):
...
+        from inline import Here
+        def m(a):
+            c = [i + 3 for i in a]
+            Here(timeout=1).given(a, [i + 1 for i in range(1000000)]).check_eq(len(c), len(a))
+    """)
+        for x in (pytester.path, checkfile):
+            items, reprec = pytester.inline_genitems(x)
+            assert len(items) == 1
+            res = pytester.runpytest()
+            assert res.ret == 0
```
Comment (sweetStreet) [inline python/tests/test_conftest.py:400]: Thanks for the test. It is better to add one failing test to check if TimeoutException is thrown out

File: python/inline/conftest.py [hunk 2]
```diff
@@ -470,10 +497,21 @@ def parse_constructor(self, node):
...
+                        and isinstance(arg, ast.Num)
+                        and isinstance(arg.value, int)
                     ):
-                        self.cur_inline_test.disabled = arg.value
+                        self.cur_inline_test.timeout = arg.value
+                    # check if "timeout" is a positive integer
+                    elif (
+                        index == 5
+                        and isinstance(arg, ast.Num)
+                        and isinstance(arg.n, int)
```
Comment (sweetStreet) [inline python/inline/conftest.py:508]: It is better to use type "float" than "int" for timeout value

File: python/inline/conftest.py [hunk 3]
```diff
@@ -526,6 +564,17 @@ def parse_constructor(self, node):
                         and isinstance(keyword.value.value, bool)
                     ):
                         self.cur_inline_test.disabled = keyword.value.value
+                    # check if "timeout" is a positive integer
+                    elif (
+                        keyword.arg == self.arg_timeout_str
+                        and isinstance(keyword.value, ast.Num)
+                        and isinstance(keyword.value.n, int)
```
Comment (sweetStreet) [inline python/inline/conftest.py:571]: It is better to use type "float" than "int" for timeout value

**Review by sweetStreet — COMMENTED** [review]
Thanks for the update!
Using asyncio seems to work. If not, maybe you can refer to this post: https://stackoverflow.com/a/22348885/6537282
Don't forget to fix the failure of test "test_check_incorrect_group_tests"

**Comment by h-alan** [comment]
> Thanks for the update! Using asyncio seems to work. If not, maybe you can refer to this post: https://stackoverflow.com/a/22348885/6537282 Don't forget to fix the failure of test "test_check_incorrect_group_tests"

That method mentioned is stated to work in only UNIX, would this cause any issues?

File: python/inline/conftest.py [hunk 4]
```diff
@@ -1,15 +1,19 @@
...
 import time
 from pathlib import Path
 from types import ModuleType
 from typing import Any, Dict, Iterable, List, Optional, Tuple, Union
 import enum
+import multiprocessing
 import pytest
 from _pytest.pathlib import fnmatch_ex, import_path
 from pytest import Collector, Config, FixtureRequest, Parser
+import asyncio
```
[the remark this answers is not in the document]
Comment (h-alan) [reply python/inline/conftest.py]: Fixed

File: python/inline/conftest.py [hunk 5]
```diff
@@ -470,10 +497,21 @@ def parse_constructor(self, node):
...
                     elif (
                         index == 4
-                        and isinstance(arg, ast.NameConstant)
-                        and isinstance(arg.value, bool)
+                        and isinstance(arg, ast.Num)
+                        and isinstance(arg.value, int)
                     ):
-                        self.cur_inline_test.disabled = arg.value
+                        self.cur_inline_test.timeout = arg.value
+                    # check if "timeout" is a positive integer
```
[the remark this answers is not in the document]
Comment (h-alan) [reply python/inline/conftest.py]: Fixed

File: python/inline/conftest.py [hunk 6]
```diff
@@ -919,16 +968,42 @@ def _find(self, tests, obj, module, globs, seen):
...
+        #     raise TimeoutException(
+        #         f"Execution timed out while running inline test"
+        #     )
+        if test.timeout >= 0: 
+            try:
+                res = await asyncio.wait_for(exec(codeobj, test.globs), timeout=test.timeout*1.0)
+                end_time = time.time()
+                out.append(f"Test Execution time: {round(end_time - start_time, 4)} seconds")
+                if test.globs:
+                    test.globs.clear()
```
[the remark this answers is not in the document]
Comment (h-alan) [reply python/inline/conftest.py]: Fixed

File: python/inline/conftest.py [hunk 7]
```diff
@@ -369,7 +380,12 @@ def parse_constructor(self, node):
                     ):
                         self.cur_inline_test.disabled = arg.value
                     # TODO: handle timeout
-
+                    elif (
+                        index == 5
+                        and isinstance(arg, ast.Constant)
+                        and isinstance(arg.value, int)
```
[the remark this answers is not in the document]
Comment (h-alan) [reply python/inline/conftest.py]: Fixed

File: python/inline/conftest.py (hunk 2 above)
Comment (h-alan) [reply python/inline/conftest.py]: Fixed

File: python/inline/conftest.py (hunk 3 above)
Comment (h-alan) [reply python/inline/conftest.py]: Fixed

File: python/inline/conftest.py [hunk 8]
```diff
@@ -919,16 +968,42 @@ def _find(self, tests, obj, module, globs, seen):
 ## InlineTest Runner
 ######################################################################
 class InlineTestRunner:
-    def run(self, test: InlineTest, out: List) -> None:
+    def __init__(self) -> None:
+        self.queue = asyncio.Queue()
```
[the remark this answers is not in the document]
Comment (h-alan) [reply python/inline/conftest.py:971]: From what I understand of asyncio it uses that queue for the timeout functions, so we just need to define the queue.

File: python/inline/conftest.py [hunk 9]
```diff
@@ -1,15 +1,19 @@
 import ast
+import asyncio
 import copy
 import inspect
+from sqlite3 import Time
```
[the remark this answers is not in the document]
Comment (zwt299) [reply python/inline/conftest.py:5]: fixed

**Pull request closed without merging**